### PR TITLE
add decompression to packet processing

### PR DESCRIPTION
**Pull request by nicHoch**
work in progress!

### CODE REVIEW

**Comment by samaloney** [comment]
I wonder if we shouldn't flatten the packets as an intermediate step, it would avoid all that set_path, get_path, get_all etc what do you think?

File: stixcore/processing/decompression.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+def decompress(packet):
+    decompression_parameter = packet.get_decompression_parameter()
+    if not decompression_parameter:
+        return 0
+    c = 1
+    for param_name, (sn, kn, mn) in decompression_parameter.items():
+        skm = (sn if isinstance(sn, int) else packet.data.get(sn),
+               packet.data.get(kn),
+               packet.data.get(mn))
+        c += packet.data.work(param_name, work_decompress, skm)
```
Comment (nicHoch) [inline stixcore/processing/decompression.py:17]: @samaloney you might wonder about that. your decompression method could deal with all all repeated NIX in the same packet at once. but then in generates a regular (n,m) array as output what might fail if there is no regular m.

so i process each array in the list as separate call within the work method. could also be helpfull for other processing steps later.

File: stixcore/tmtc/parser.py [hunk 2]
```diff
@@ -37,10 +38,130 @@ def parameter_dict_2_PacketData(cls, d):
...
+            The NIX name of the parameter.
+
+        Returns
+        -------
+        'any'
+            The found value.
+        """
+        return getattr(self, nix)
+
+    def apply(self, nix, callback, args, addnix=None):
```
Comment (samaloney) [inline stixcore/tmtc/parser.py:132]: you could add **kwargs here and pass them to the callback too if we need to might be useful later

File: stixcore/tmtc/parser.py [hunk 3]
```diff
@@ -37,10 +38,130 @@ def parameter_dict_2_PacketData(cls, d):
...
+        """Apply a processing method to a parameter.
+
+        Overids the value after processing or creates a new one.
+
+        Parameters
+        ----------
+         nix : `str`
+            The NIX name of the parameter.
+        callback : function
```
Comment (samaloney) [inline stixcore/tmtc/parser.py:141]: Could be a class either - 'callable' might be more descriptive

**Review by samaloney — APPROVED** [review]
Looks good to me - a few comment but nothing that _has_ to be changed.

**Pull request merged**